src/main/kotlin/com/natpryce/await/AwaitableResult.kt

```diff
@@ -0,0 +1,31 @@
+package com.natpryce.await
+
+import com.natpryce.Failure
+import com.natpryce.Result
+import com.natpryce.Success
+import com.natpryce.onFailure
+import com.natpryce.await.NoStackTraceException.AwaitFailure
+
+class AwaitableResult<E> private constructor() {
+
+    private lateinit var failure: Failure<E>
+
+    fun <T> Result<T, E>.await(): T {
+        return this.onFailure {
+            failure = it
+            throw AwaitFailure
+        }
+    }
+
+    companion object {
+        fun <T, E> with(successBlock: AwaitableResult<E>.() -> T): Result<T, E> {
```

Comment:
marking this as `inline` significantly improved my benchmark test (by about 100000 ops/ms for both success and failure flow) but from what I had read about the performance cost that could be incurred for a potentially large `successBlock` I wasn't sure it was worth it. Wasn't sure what would be a good way to even benchmark this so leaving as is. 

The other disadvantage in its current form is `AwaitableResult` constructor would have to be made public which I think would be confusing for a user on how to use the dsl. That said could make this a separate object instead of a companion and that would at least resolve that issue - maybe name the class `AwaitableResultContext` instead

Reply:
see latest comment